File: sample-chat/src/main/java/com/quickblox/sample/chat/utils/qb/callback/QbEntityCallbackTwoTypeWrapper.java
package com.quickblox.sample.chat.utils.qb.callback;

import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;

import com.quickblox.core.QBEntityCallback;
import com.quickblox.core.exception.QBResponseException;

public class QbEntityCallbackTwoTypeWrapper<T, R> implements QBEntityCallback<T> {
    protected static final Handler mainThreadHandler = new Handler(Looper.getMainLooper());
    protected QBEntityCallback<R> callback;

    public QbEntityCallbackTwoTypeWrapper(QBEntityCallback<R> callback) {
        this.callback = callback;
    }

    @Override
    public void onSuccess(T t, Bundle bundle) {
        // Do nothing, we want to trigger callback with another data type
    }

    @Override
    public void onError(QBResponseException error) {
        onErrorInMainThread(error);
    }

    protected void onSuccessInMainThread(final R result, final Bundle bundle) {
        mainThreadHandler.post(new Runnable() {
            @Override
            public void run() {
                callback.onSuccess(result, bundle);
            }
        });
    }

    protected void onErrorInMainThread(final QBResponseException error) {
        mainThreadHandler.post(new Runnable() {
            @Override
            public void run() {
                callback.onError(error);
            }
        });
    }
}

File: sample-chat-kotlin/app/src/main/java/com/quickblox/sample/chat/kotlin/utils/qb/callback/QbEntityCallbackTwoTypeWrapper.kt
package com.quickblox.sample.chat.kotlin.utils.qb.callback

import android.os.Bundle
import android.os.Handler
import android.os.Looper
import com.quickblox.core.QBEntityCallback
import com.quickblox.core.exception.QBResponseException

open class QbEntityCallbackTwoTypeWrapper<T, R>(private var callback: QBEntityCallback<R>) : QBEntityCallback<T> {
    private val mainThreadHandler = Handler(Looper.getMainLooper())

    override fun onSuccess(t: T, bundle: Bundle?) {
        // Do nothing, we want to trigger callback with another data type
    }

    override fun onError(error: QBResponseException) {
        onErrorInMainThread(error)
    }

    protected fun onSuccessInMainThread(result: R, bundle: Bundle?) {
        mainThreadHandler.post { callback.onSuccess(result, bundle) }
    }

    private fun onErrorInMainThread(error: QBResponseException) {
        mainThreadHandler.post { callback.onError(error) }
    }
}